multi-GPU simulations broken for dust build

After commit a94a4d2d on dev I am no longer able to run a simulation with the dust build on more than one GPU. I tried compiling both the dust and hydro build checked out to this commit and got the following error:

```
[her45@ppc-n0 cholla]$ make TYPE=hydro -j
builds/prereq.sh build crc
mpicc -Ofast -DCUDA -DMPI_CHOLLA -DBLOCK -DPRECISION=2 -DPPMC -DHLLC -DAVERAGE_SLOW_CELLS -DTEMPERATURE_FLOOR -DVL -DSCALAR -DDUST -DCOOLING_GPU -DSLICES -DPROJECTION -DOUTPUT -DHDF5 -DMPI_GPU -DGIT_HASH='"a94a4d2d0556407b94021a6dbb2e18b9915c8e45"' -DMACRO_FLAGS='"-DCUDA -DMPI_CHOLLA -DBLOCK -DPRECISION=2 -DPPMC -DHLLC -DAVERAGE_SLOW_CELLS -DTEMPERATURE_FLOOR -DVL -DSCALAR -DDUST -DCOOLING_GPU -DSLICES -DPROJECTION -DOUTPUT -DHDF5 -DMPI_GPU -DGIT_HASH='"a94a4d2d0556407b94021a6dbb2e18b9915c8e45"'"' -Isrc -c src/mpi/MPI_Comm_node.c -o src/mpi/MPI_Comm_node.o
mpicxx -Ofast -std=c++17 -DCUDA -DMPI_CHOLLA -DBLOCK -DPRECISION=2 -DPPMC -DHLLC -DAVERAGE_SLOW_CELLS -DTEMPERATURE_FLOOR -DVL -DSCALAR -DDUST -DCOOLING_GPU -DSLICES -DPROJECTION -DOUTPUT -DHDF5 -DMPI_GPU -DGIT_HASH='"a94a4d2d0556407b94021a6dbb2e18b9915c8e45"' -DMACRO_FLAGS='"-DCUDA -DMPI_CHOLLA -DBLOCK -DPRECISION=2 -DPPMC -DHLLC -DAVERAGE_SLOW_CELLS -DTEMPERATURE_FLOOR -DVL -DSCALAR -DDUST -DCOOLING_GPU -DSLICES -DPROJECTION -DOUTPUT -DHDF5 -DMPI_GPU -DGIT_HASH='"a94a4d2d0556407b94021a6dbb2e18b9915c8e45"'"' -Isrc -I/ihome/crc/install/power9/hdf5/1.12.0/build-gcc-10.1.0/include -I/ihome/crc/install/power9/cuda/11.1.0/include -c src/analysis/analysis.cpp -o src/analysis/analysis.o
src/mpi/MPI_Comm_node.c:2:12: fatal error: ../mpi/MPI_Comm_node.h: No such file or directory
    2 |   #include "../mpi/MPI_Comm_node.h"
      |            ^~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make: *** [Makefile:206: src/mpi/MPI_Comm_node.o] Error 1
make: *** Waiting for unfinished jobs....

```

The next commit after this (4a92255) does compile for both dust and hydro builds, but when I actually run the simulation using the command `mpirun -np 2 ./cholla.dust.crc cloud-wind.txt` I get the following error:

```
[her45@ppc-n0 2023-03-10]$ mpirun -np 2 ./cholla.dust.crc cloud-wind.txt 
--------------------------------------------------------------------------
By default, for Open MPI 4.0 and later, infiniband ports on a device
are not used by default.  The intent is to use UCX for these devices.
You can override this policy by setting the btl_openib_allow_ib MCA parameter
to true.

  Local host:              ppc-n0
  Local adapter:           mlx5_0
  Local port:              1

--------------------------------------------------------------------------
--------------------------------------------------------------------------
WARNING: There was an error initializing an OpenFabrics device.

  Local host:   ppc-n0
  Local device: mlx5_0
--------------------------------------------------------------------------
Git Commit Hash = 4a92255f98962ac2be1bf7eeeb5111042d096437
Macro Flags     = -DCUDA -DMPI_CHOLLA -DBLOCK -DPRECISION=2 -DPPMC -DHLLC -DAVERAGE_SLOW_CELLS -DTEMPERATURE_FLOOR -DVL -DSCALAR -DDUST -DCOOLING_GPU -DSLICES -DPROJECTION -DOUTPUT -DHDF5 -DMPI_GPU -DGIT_HASH=4a92255f98962ac2be1bf7eeeb5111042d096437
Parameter values:  nx = 1024, ny = 256, nz = 256, tout = 60000.000000, init = Clouds, boundaries = 4 3 3 3 3 3
Output directory:  ./hdf5/

Creating Log File: run_output.log 

  File exists, appending values: run_output.log 

nproc_x 2 nproc_y 1 nproc_z 1
Allocating MPI communication buffers on GPU (nx = 1572864, ny = 3194880, nz = 3294720).
Cloud positions: 0.080000 0.020000 0.020000
Local number of grid cells: 512 256 256 36241920
Setting initial conditions...
Cloud positions: 0.080000 0.020000 0.020000
Initial conditions set.
Setting boundary conditions...
Boundary conditions set.
Dimensions of each cell: dx = 0.000156 dy = 0.000156 dz = 0.000156
Ratio of specific heats gamma = 1.666667
Nstep = 0  Simulation time = 0.000000
Writing initial conditions to file...

Saving Snapshot: 0 
[ppc-n0.crc.pitt.edu:119982] 1 more process has sent help message help-mpi-btl-openib.txt / ib port not selected
[ppc-n0.crc.pitt.edu:119982] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
[ppc-n0.crc.pitt.edu:119982] 1 more process has sent help message help-mpi-btl-openib.txt / error in device init
Starting calculations.
CUDA ERROR AT LINE 37 OF FILE 'src/grid/cuda_boundaries.cu': cudaErrorIllegalAddress an illegal memory access was encountered
CUDA ERROR AT LINE 37 OF FILE 'src/grid/cuda_boundaries.cu': cudaErrorIllegalAddress an illegal memory access was encountered
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [[60223,1],1]
  Exit code:    188
```

Note that I have no problem running a smaller version of this simulation that fits on one GPU with this version of Cholla. I am able to run it seemingly without any issues on one GPU with just the executable and the `mpirun` command. I'm also able to get it to "run" on one GPU when I run the larger simulation with `mpirun` (although nothing is actually run because the GPU is overloaded). The problem occurs only when I specify that I want to use 2 GPUs with `mpirun`.

--- a/src/mpi/mpi_routines.cpp
+++ b/src/mpi/mpi_routines.cpp
@@ -5,6 +5,7 @@
   #include <mpi.h>
 
   #include <iostream>
+  #include <tuple>
 
   #include "../global/global.h"
   #include "../io/io.h"
@@ -213,6 +214,9 @@ void InitializeChollaMPI(int *pargc, char **pargv[])
     chexit(-2);
   }
   #endif
+
+  /*set up node communicator*/
+  std::tie(procID_node, nproc_node) = MPI_Comm_node();
 
   // #ifdef ONLY_PARTICLES
   // chprintf("ONLY_PARTICLES: Initializing without CUDA support.\n");
@@ -976,6 +980,40 @@ void copyHostToDeviceReceiveBuffer(int direction)
       cudaMemcpy(d_recv_buffer_z1, h_recv_buffer_z1, zbsize * sizeof(Real), cudaMemcpyHostToDevice);
       break;
   }
+}
+
+std::pair<int, int> MPI_Comm_node()
+{
+  // get the global process rank
+  int myid, nproc;
+  MPI_Comm_rank(MPI_COMM_WORLD, &myid);
+  MPI_Comm_size(MPI_COMM_WORLD, &nproc);
+
+  // if there is the only one process, then just return the global rank and size
+  if (nproc == 1) {
+    return {myid, nproc};
+  }
+
+  // get the hostname of the node
+  std::string pname;  // node hostname
+  pname.resize(MPI_MAX_PROCESSOR_NAME);
+  int pname_length;  // length of node hostname
+
+  MPI_Get_processor_name(pname.data(), &pname_length);
+
+  // hash the name of the node. MPI_Comm_split doesn't like negative numbers and accepts ints not unsigned ints so we
+  // need to take the absolute value
+  int const hash = std::abs(static_cast<int>(std::hash<std::string>{}(pname)));
+
+  // split the communicator
+  MPI_Comm node_comm;  // communicator for the procs on each node
+  MPI_Comm_split(MPI_COMM_WORLD, hash, myid, &node_comm);
+
+  // get size and rank
+  MPI_Comm_rank(node_comm, &myid);
+  MPI_Comm_size(node_comm, &nproc);
+
+  return {myid, nproc};
 }
 
 #endif /*MPI_CHOLLA*/

--- a/src/mpi/mpi_routines.h
+++ b/src/mpi/mpi_routines.h
@@ -4,6 +4,8 @@
     #include <mpi.h>
     #include <stddef.h>
 
+    #include <utility>
+
     #include "../global/global.h"
     #include "../grid/grid3D.h"
 
@@ -198,6 +200,13 @@ void deallocate_three_dimensional_int_array(int ***x, int n, int l, int m);
 
 /* Copy MPI receive buffers on Host to their device locations */
 void copyHostToDeviceReceiveBuffer(int direction);
+
+/*!
+ * \brief Split the communicator for each node and return IDs
+ *
+ * \return std::pair<int, int> The rank id and total number of processes
+ */
+std::pair<int, int> MPI_Comm_node();
 
   #endif /*MPI_ROUTINES_H*/
 #endif   /*MPI_CHOLLA*/

